replay_provider.go

```diff
@@ -1,48 +1,136 @@
 // The events must have an ID unless the AutoIDs flag is toggled.
 type FiniteReplayProvider struct {
-	b buffer
-
-	// Count is the maximum number of events FiniteReplayProvider should hold as valid.
-	// It must be a positive integer, or the code will panic.
-	Count int
-	// AutoIDs configures FiniteReplayProvider to automatically set the IDs of events.
-	AutoIDs bool
+	mut       sync.RWMutex
+	cap       int
+	buf       []messageWithTopics
+	head      int
+	tail      int
+	count     int
+	autoIDs   bool
+	currentID int64
 }
 
 // Put puts a message into the provider's buffer. If there are more messages than the maximum
 // number, the oldest message is removed.
 func (f *FiniteReplayProvider) Put(message *Message, topics []string) *Message {
-	if f.b == nil {
-		f.b = getBuffer(f.AutoIDs, f.Count)
+	if len(topics) == 0 {
+		panic(errors.New(
+			"go-sse: no topics provided for Message.\n" +
+				formatMessagePanicString(message)))
 	}
 
-	if f.b.len() == f.b.cap() {
-		f.b.dequeue()
+	f.mut.Lock()
+	defer f.mut.Unlock()
+
+	if f.autoIDs {
+		f.currentID++
+
+		message.ID = ID(strconv.FormatInt(f.currentID, 10))
+	} else if !message.ID.IsSet() {
+		panicString := "go-sse: a Message without an ID was given to a provider that doesn't set IDs automatically.\n" + formatMessagePanicString(message)
+
+		panic(errors.New(panicString))
+	}
+
+	f.buf[f.tail] = messageWithTopics{message: message, topics: topics}
+
+	f.tail++
+	if f.tail >= f.cap {
+		f.tail = 0
 	}
 
-	return f.b.queue(message, topics)
+	if f.tail == f.head {
+		f.head = f.tail + 1
+
+		if f.head > f.cap {
+			f.head = 0
+		}
+	}
+
+	return message
 }
 
 // Replay replays the messages in the buffer to the listener.
 // It doesn't take into account the messages' expiry times.
 func (f *FiniteReplayProvider) Replay(subscription Subscription) error {
-	if f.b == nil {
+	f.mut.RLock()
+
+	if f.head == f.tail {
+		f.mut.RUnlock()
 		return nil
 	}
 
-	events := f.b.slice(subscription.LastEventID)
-	if len(events) == 0 {
-		return nil
+	// The assumption here is that most of the lifetime of an application
+	// the circular buffer would be full, so we might as well allocate in
+	// one go.
+	vs := make([]messageWithTopics, 0, f.cap)
+
+	var n int
+
+	// Copy head to end and start to tail when head is after tail.
+	if f.tail < f.head {
+		n = f.cap - f.tail
+		copy(vs[0:n], f.buf[f.tail:])
+		copy(vs[n:n+f.tail], f.buf[0:f.tail])
+		// If the head is after the tail the buffer is full.
+		n = f.cap
+	} else {
+		n = f.tail - f.head
+		copy(vs[0:n], f.buf[f.head:f.tail])
+	}
+
+	f.mut.RUnlock()
+
+	values := vs[0:n]
+
+	for i := range values {
+		// This could be done as part of the the initial copy to vs,
+		// leaving it as a separate op for now.
+		if values[i].message.ID.value == subscription.LastEventID.value {
+			if i == len(values)-1 {
+				values = nil
+			} else {
+				values = values[i+1:]
+			}
+
+			break
+		}
```

Comment:
From what I understand, this part puts the events from the buffer in the order in which they should be replayed. Couldn't we do it without any allocations?

I imagine we could iterate over:
1. `f.buf[f.head:f.tail]`, when `f.head  < f.tail`
1. `f.buf[f.head:]` and `f.buf[:f.tail]` in this order, when `f.tail < f.head`

We could have an internal function which loops over a slice, finds the first event with the ID equal to the given `LastEventID` and starts replaying from there. Something like:

```go
func replay(sub Subscription, events []messageWithTopics, foundFirstEvent bool) (hasFoundFirstEvent bool, err error) {
    for _, e := range events {
        if !foundFirstEvent && e.message.ID == sub.LastEventID {
            foundFirstEvent = true
        }
        if foundFirstEvent && topicsIntersect(subscription.Topics, e.Topics) {
            if err := subscription.Client.Send(e.message); err != nil {
                return false, err
            }
        }
    }
    return foundFirstEvent, nil
}
```
and it would be used like (example for scenario 2):
```go
foundFirst, err := replay(subscription, f.buf[f.head:], false)
if err != nil {
    return err
}
_, err := replay(subscription, f.buf[:f.tail], foundFirst)
return err
```

What do you think?

Reply:
Yeah, that looks like a good solution. I was a bit worried about holding the lock while doing the scanning and filtering as I assumed concurrent operations, but I see now that Joe uses an eventloop model to processing so it's serialised at the root.